sequence_processing_pipeline/Pipeline.py

```diff
@@ -276,6 +297,69 @@ def _validate_sample_sheet(self, sample_sheet_path):
                               isinstance(x, WarningMessage)]
             return val_sheet
 
+    def _validate_mapping_file(self, mapping_file_path):
+        """
+        Performs validation for mapping-files.
+        :return: If successful, a valid mapping-file. Raises descriptive
+                 PipelineError() on all failures. Warning messages are
+                 appended to self.warnings.
+        """
+        try:
+            df = pd.read_csv(mapping_file_path, delimiter='\t', dtype=str)
+        except pd.errors.ParserError:
+            # ignore parser errors as they obviously prove this is not a
+            # valid mapping file. Return None w/error message
+            raise PipelineError('Cannot parse mapping-file.')
+
+        # first, detect any duplicate column names, regardless of any mixed-
+        # capitalization, and notify the user.
+        d = defaultdict(list)
+        for column in df.columns:
+            d[column.lower()].append(column)
+
+        # generate a list of all unique column names that appear more than
+        # once, regardless of capitalization.
+        duplicate_columns = [col for col in d.keys() if len(d[col]) > 1]
+
+        # generate a list containing lists of duplicate column names in their
+        # original case to report to the user.
+        dupes = [d[column] for column in duplicate_columns]
+
+        if dupes:
+            # column-names are case-insensitive, and must be unique.
+            # return groups of duplicate column names (differentiated only by
+            # a different mixed-case) to the user.
+            raise PipelineError("Mapping-file contains duplicate columns: "
+                                "%s" % ', '.join([str(tpl) for tpl in dupes]))
```

Comment:
In this code, we are actually discarding the original values and just displaying the lower cased duplicated value; for example, if we had 2 columns: `cAse`, `Case`, we are only displaying `case`, why not use `from collections import Counter` and do something like:
`dupes = [k for k, v in Counter([c.lower() for c in df.columns]).items() if v > 1)`

Reply:
Change based on code below:
```
columns = Counter([c.lower() for c in df.columns])
dupes = [k for k, v in columns.items() if v > 1)
```

Reply:
Actually, d is a dictionary with the lower()ed version of the column name as key and all mixed case variations of that column-name as a list of values. This is actually the metadata we want to return to the user.

The issue with your proposed changes is that for a list of column names:
l = ['a', 'A', 'b', 'B', 'C']

Your new code will generate:
['a', 'b']

instead of:
[['a', 'A'], ['b', 'B']]

I think the use of Counter() in this case won't improve things as we need the list of variations, not just their number of occurrences.

Reply:
That's my point, I don't think we need the variations as they are not being used anywhere, right?

Reply:
https://github.com/charles-cowart/mg-scripts/blob/f0af5a55d738e1870da59e54de75a7344bae3dfb/sequence_processing_pipeline/Pipeline.py#L333